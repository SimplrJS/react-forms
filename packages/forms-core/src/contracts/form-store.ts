import { Dictionary, JsonValue, NestedDictionary } from "./helpers";
import { Draft } from "immer";

export interface FieldData<TValue = unknown, TRenderValue = unknown> {
    currentValue: TValue;
    initialValue: TValue;
    defaultValue: TValue;
    transientValue?: TRenderValue;
}

export interface FormStatus {
    focused: boolean;
    touched: boolean;
    pristine: boolean;
    disabled: boolean;
    readonly: boolean;
}

// TODO: THydrationValue example with DraftJS: TValue = DraftJsState, THydrationValue = object.
export interface FieldState<TData extends object, THydrationValue extends JsonValue = JsonValue> {
    id: string;
    name: string;

    dehydrate: (state: this) => THydrationValue;
    hydrate: (value: THydrationValue) => FieldState<TData, THydrationValue>;

    status: FormStatus;
    data: TData;

    fields: Dictionary<FieldState<TData, THydrationValue>>;
}

export interface FormStoreData {
    submitCallback?: () => void;
    dehydratedState: NestedDictionary<unknown>;
}

export interface FormState<TData extends object = FormStoreData> extends FieldState<TData> {}

export type FormStateUpdater<TBase> = (draft: Draft<TBase>) => void;
