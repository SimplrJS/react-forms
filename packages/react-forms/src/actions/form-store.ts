// TODO: extract into abstractions
// This requires tooling update, because webpack goes crazy now
export abstract class FormAction {
    constructor(protected formId: string) { }

    public get FormId(): string {
        return this.formId;
    }
}

export abstract class FieldAction extends FormAction {
    constructor(formId: string, protected fieldId: string) {
        super(formId);
    }

    public get FieldId(): string {
        return this.fieldId;
    }
}

export class StateChanged extends FormAction { }

export class FieldRegistered extends FieldAction { }

export class FieldsGroupRegistered extends FormAction {
    constructor(protected formId: string, private fieldsGroupId: string) {
        super(formId);
    }

    public get FieldsGroupId(): string {
        return this.fieldsGroupId;
    }
}

export class FieldsArrayRegistered extends FormAction {
    constructor(protected formId: string, private fieldsGroupId: string) {
        super(formId);
    }

    public get FieldsArrayId(): string {
        return this.fieldsGroupId;
    }
}

export class ValueChanged extends FieldAction { }

export class FieldPropsChanged extends FieldAction { }

export class FormPropsChanged extends FormAction { }

export class FormDisabled extends FormAction { }

export class FormEnabled extends FormAction { }

export class FieldTouched extends FieldAction { }

export class FieldValidated extends FieldAction { }

export class FieldActive extends FormAction {
    constructor(formId: string, private fieldId: string | undefined) {
        super(formId);
    }

    public get FieldId(): string | undefined {
        return this.fieldId;
    }
}

export class FieldBlurred extends FieldAction { }
