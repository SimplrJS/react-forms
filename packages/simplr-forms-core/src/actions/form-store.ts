export class StateUpdated { }

export class ValueChanged {
    constructor(private fieldId: string, private newValue: any) { }

    public get FieldId() {
        return this.fieldId;
    }

    public get NewValue() {
        return this.newValue;
    }
}

export class PropsChanged {
    constructor(private fieldId: string) { }

    public get FieldId() {
        return this.fieldId;
    }
}
