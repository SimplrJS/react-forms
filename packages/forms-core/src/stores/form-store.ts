import produce, { Draft } from "immer";
import { TinyEmitter } from "@reactway/tiny-emitter";
import { FormState } from "../contracts/form-store";
import { NestedDictionary } from "../contracts/helpers";
import { dehydrateField, hydrateField } from "./form-store-helpers";

export class FormStore extends TinyEmitter {
    protected formId: string;
    protected state: FormState;
    protected dehydratedState: NestedDictionary<unknown> = {};

    constructor() {
        super();
        this.formId = "";
        this.state = this.getInitialState(this.formId);
    }

    public getState(): FormState {
        return this.state;
    }

    public update(updater: (draft: Draft<FormState>) => void): void {
        const newState = produce(this.state, draft => {
            return updater(draft);
        });

        if (this.state === newState) {
            return;
        }
        this.state = newState;

        this.emit();
    }

    protected getInitialState(id: string): FormState {
        const initialState: FormState = {
            id: id,
            name: id,
            data: {
                dehydratedState: {}
            },
            status: {
                focused: false,
                pristine: true,
                touched: false,
                disabled: false,
                readonly: false
            },
            fields: {},
            dehydrate: dehydrateField,
            hydrate: hydrateField
        };

        return produce<FormState>(initialState, () => {});
    }
}
