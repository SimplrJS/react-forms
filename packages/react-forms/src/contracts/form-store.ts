import * as Immutable from "immutable";
import { TypedRecord } from "typed-immutable-record";

import { FieldStoreStateRecord } from "../contracts/field";
import { FieldsGroupStoreStateRecord } from "../contracts/fields-group";
import { FormStateRecord } from "../contracts/form";
import { FieldValidationStatus } from "./validation";

export interface FormStoreState extends FormStoreStateStatus {
    Fields: Immutable.Map<string, FieldStoreStateRecord>;
    FieldsGroups: Immutable.Map<string, FieldsGroupStoreStateRecord>;
    FieldsValidationStatuses: Immutable.Map<string, FieldValidationStatus>;
    Form: FormStateRecord;
}

export interface FormStoreStateStatus {
    Validating: boolean;
    Submitting: boolean;
    Disabled: boolean;
    HasError: boolean;
    Pristine: boolean;
    Touched: boolean;
}

export interface FormStoreStateRecord extends TypedRecord<FormStoreStateRecord>, FormStoreState { }

export interface BuiltFormObject {
    Fields: Immutable.Map<string, FieldStoreStateRecord>;
    Object: any;
}
