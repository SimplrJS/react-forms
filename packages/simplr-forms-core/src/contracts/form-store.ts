import * as Immutable from "immutable";
import { TypedRecord } from "typed-immutable-record";

import { FieldStateRecord } from "../contracts/field";
import { FieldsGroupStoreState } from "../contracts/fields-group";
import { FormStateRecord } from "../contracts/form";

export interface FormStoreState extends FormStoreStateProperties {
    Fields: Immutable.Map<string, FieldStateRecord>;
    FieldsGroups: Immutable.Map<string, FieldsGroupStoreState>;
    Form: FormStateRecord;
}

// TODO: Naming
export interface FormStoreStateProperties {
    Validating: boolean;
    HasError: boolean;
    Pristine: boolean;
    Touched: boolean;
}

export interface FormStoreStateRecord extends TypedRecord<FormStoreStateRecord>, FormStoreState { }

export interface BuiltFormObject {
    Fields: Immutable.Map<string, FieldStateRecord>;
    Object: any;
}
