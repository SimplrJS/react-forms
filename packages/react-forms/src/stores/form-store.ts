import { BaseStore } from "../abstractions/base-store";
import { FieldStore, FieldStoreData } from "./field-store";

import { StoreHydrated } from "../actions/store-actions";
import { FieldRegistered, FieldUnregistered } from "../actions/field-actions";

export interface FormStoreState {
    fields: { [key: string]: FieldStore };
}

export interface FormStoreData {
    fields: { [key: string]: FieldStoreData };
}

export class FormStore extends BaseStore<FormStoreState, FormStoreData> {
    constructor(protected readonly formId: string) {
        super();
    }

    protected getInitialState(): FormStoreState {
        return {
            fields: {}
        };
    }

    public registerField(fieldId: string): void {
        if (this.getState().fields[fieldId] != null) {
            throw Error(`@simplr/react-forms: Field '${fieldId}' in '${this.formId}' already exists.`);
        }

        const storeInstance: FieldStore = new FieldStore(this.formId, fieldId);

        this.setState(new FieldRegistered(fieldId), state => ({
            fields: {
                ...state.fields,
                [`${fieldId}`]: storeInstance
            }
        }));
    }

    public unregisterField(fieldId: string): void {
        this.setState(new FieldUnregistered(fieldId), state => {
            const form = this.getState().fields[fieldId];
            if (form == null) {
                return state;
            }

            const { [`${fieldId}`]: deletedField, ...restFields } = state.fields;

            return {
                ...state,
                fields: restFields
            };
        });
    }

    public getField(fieldId: string): FieldStore | undefined {
        return this.getState().fields[fieldId];
    }

    public getFields(): { [key: string]: FieldStore | undefined } {
        return this.getState().fields;
    }

    public hydrate(data: FormStoreData): void {
        const state = this.getInitialState();

        for (const fieldId in data.fields) {
            if (data.fields.hasOwnProperty(fieldId)) {
                const field = new FieldStore(this.formId, fieldId);
                field.hydrate(data.fields[fieldId]);

                state.fields[fieldId] = field;
            }
        }

        this.emit(new StoreHydrated());
    }

    public dehydrate(): FormStoreData {
        const state = this.getState();
        const data: FormStoreData = {
            fields: {}
        };

        for (const fieldId in state.fields) {
            if (state.fields.hasOwnProperty(fieldId)) {
                data.fields[fieldId] = state.fields[fieldId].dehydrate();
            }
        }

        return data;
    }
}
