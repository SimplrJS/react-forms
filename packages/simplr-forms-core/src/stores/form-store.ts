import * as Immutable from "immutable";
import { recordify } from "typed-immutable-record";
import { ActionEmitter } from "action-emitter";

import * as Actions from "../actions/form-store";
import {
    FieldState,
    FieldValue,
    FieldStateRecord,
    FormErrorRecord,
    FieldProps,
    FieldStatePropsRecord,
    FieldStateProps
} from "../contracts/field";
import { FormState, FormStateRecord } from "../contracts/form";
import { FormStoreState, FormStoreStateRecord } from "../contracts/form-store";
import { FieldsGroupStateRecord } from "../contracts/fields-group";
import { ConstructFormError } from "../utils/form-error-helpers";
import { FormError } from "../contracts/error";

export class FormStore extends ActionEmitter {
    constructor(formId: string) {
        super();
        this.FormId = formId;
        this.state = this.GetInitialFormStoreState();
    }

    protected FormId: string;

    private state: FormStoreStateRecord;
    protected get State(): FormStoreStateRecord {
        return this.state;
    }
    protected set State(newState: FormStoreStateRecord) {
        this.state = newState;

        this.emit(new Actions.StateUpdated());
    }

    public GetState(): FormStoreStateRecord {
        return this.State;
    }

    /**
     * ========================
     *  Public API
     * ========================
     */

    /**
     * Constructs field id from given fieldName and an optional fieldsGroupIdkds
     *
     * @param {string} fieldName
     * @param {string} [fieldsGroupId]
     * @returns Constructed field id
     *
     * @memberOf FormStore
     */
    public GetFieldId(fieldName: string, fieldsGroupId?: string) {
        if (fieldsGroupId != null) {
            return `${fieldsGroupId}___${fieldName}`;
        }

        return fieldName;
    }

    public RegisterField(
        fieldId: string,
        initialValue: FieldValue,
        defaultValue: FieldValue,
        props?: FieldStateProps,
        fieldsGroupId?: string
    ) {
        // Construct field state
        let fieldState = this.GetInitialFieldState();
        fieldState.DefaultValue = defaultValue;
        fieldState.Value = initialValue;

        if (props != null) {
            fieldState.Props = recordify<FieldStateProps, FieldStatePropsRecord>(props);
        }

        if (fieldsGroupId != null) {
            fieldState.FieldsGroup = {
                Id: fieldsGroupId
            };
        }

        // Add field into form store state
        this.State = this.State.withMutations(state => {
            state.Fields = state.Fields.set(fieldId, recordify<FieldState, FieldStateRecord>(fieldState));
        });

        this.emit(new Actions.FieldRegistered(fieldId, initialValue));
    }

    public UnregisterField(fieldId: string) {
        // Remove field from form store state
        this.State = this.State.withMutations(state => {
            state.Fields = state.Fields.remove(fieldId);
        });
    }

    public HasField(fieldId: string): boolean {
        return this.State.Fields.has(fieldId);
    }

    public GetField(fieldId: string): FieldStateRecord {
        return this.State.Fields.get(fieldId);
    }

    public SetSubmitCallback(submitCallback: () => void) {
        this.State = this.State.withMutations(state => {
            state.Form = state.Form.merge({
                SubmitCallback: submitCallback
            } as FormState);
        });
    }

    public UpdateProps(fieldId: string, props: FieldStateProps) {
        const propsRecord = recordify<FieldStateProps, FieldStatePropsRecord>(props);
        const fieldState = this.State.Fields.get(fieldId);

        if (fieldState.Props == null || fieldState.Props.equals(propsRecord)) {
            return;
        }

        this.State = this.State.withMutations(state => {
            const fieldState = state.Fields.get(fieldId);
            state.Fields = state.Fields.set(fieldId, fieldState.merge({
                Props: recordify<FieldStateProps, FieldStatePropsRecord>(props)
            } as FieldState));
        });

        this.emit(new Actions.PropsChanged(fieldId));
    }

    public ValueChanged(fieldId: string, newValue: FieldValue) {
        this.State = this.State.withMutations(state => {
            const fieldState = state.Fields.get(fieldId);
            state.Fields = state.Fields.set(fieldId, fieldState.merge({
                Value: newValue
            } as FieldState));
        });

        this.emit(new Actions.ValueChanged(fieldId, newValue));
    }

    public async Validate(fieldId: string, validationPromise: Promise<void>) {
        const field = this.State.Fields.get(fieldId);
        const fieldValue = field.Value;

        // Skip if it's already validating
        if (!field.Validating) {
            this.State = this.State.withMutations(state => {
                // Set form state to Validating: true
                state.Form = state.Form.merge({
                    Validating: true,
                    Error: undefined
                } as FormState);

                const fieldState = state.Fields.get(fieldId);
                state.Fields = state.Fields.set(fieldId, fieldState.merge({
                    Validating: true,
                    Error: undefined
                } as FieldState));
            });
        }

        try {
            // Wait for validation to finish
            await validationPromise;

            // Skip validation if the value has changed again
            const currentFieldValue = this.State.Fields.get(fieldId).Value;
            if (currentFieldValue !== fieldValue) {
                return;
            }

            this.State = this.State.withMutations(state => {
                const fieldState = state.Fields.get(fieldId);
                state.Fields = state.Fields.set(fieldId, fieldState.merge({
                    Validating: false
                } as FieldState));
            });
        } catch (error) {
            // Skip validation if the value has changed again
            const currentFieldValue = this.State.Fields.get(fieldId).Value;
            if (currentFieldValue !== fieldValue) {
                return;
            }

            const formError = ConstructFormError(error);
            if (formError == null) {
                throw Error(error);
            }

            this.State = this.State.withMutations(state => {
                const fieldState = state.Fields.get(fieldId);
                state.Fields = state.Fields.set(fieldId, fieldState.merge({
                    Validating: false,
                    Error: recordify<FormError, FormErrorRecord>(formError!)
                } as FieldState));
            });
        }
    }

    public InitiateSubmit() {
        if (this.State.Form.SubmitCallback == null) {
            throw new Error("simplr-forms-core: Submit method is called before SubmitCallback is set.");
        }
        this.State.Form.SubmitCallback();
    }

    // public Submit(submitPromise: Promise<>) {

    // }

    /**
     * ========================
     *  Local helper methods
     * ========================
     */

    protected GetInitialFormStoreState(): FormStoreStateRecord {
        return recordify<FormStoreState, FormStoreStateRecord>({
            Fields: Immutable.Map<string, FieldStateRecord>(),
            FieldsGroups: Immutable.Map<string, FieldsGroupStateRecord>(),
            Form: recordify<FormState, FormStateRecord>(this.GetInitialFormState())
        });
    }

    protected GetInitialFormState(): FormState {
        return {
            Validating: false,
            Submitting: false,
            Pristine: true,
            SuccessfullySubmitted: false,
            ActiveFieldId: undefined,
            Error: undefined,
            SubmitCallback: undefined
        };
    }

    protected GetInitialFieldState(): FieldState {
        return {
            DefaultValue: undefined,
            Value: undefined,
            Touched: false,
            Pristine: true,
            Validating: false,
            Error: undefined,
            FieldsGroup: undefined,
            Props: undefined
        };
    }
}
