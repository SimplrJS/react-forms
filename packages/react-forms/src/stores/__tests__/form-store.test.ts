import { FormStore } from "../form-store";

const FAKE_FORM_ID = "fake-form-id";
const FAKE_FIELD_ID = "fake-field-id";

it("Register a field", () => {
    const store = new FormStore(FAKE_FORM_ID);
    store.registerField(FAKE_FIELD_ID);

    expect(store.getFields()).toMatchSnapshot();
});

it("Dehydration", () => {
    const store = new FormStore(FAKE_FORM_ID);
    store.registerField(FAKE_FIELD_ID);

    expect(store.dehydrate()).toMatchSnapshot();
});
