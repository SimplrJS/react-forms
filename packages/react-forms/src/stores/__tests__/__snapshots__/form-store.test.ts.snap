// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dehydration 1`] = `
Object {
  "fields": Object {
    "fake-field-id": Object {
      "value": "",
    },
  },
}
`;

exports[`Register a field 1`] = `
Object {
  "fake-field-id": FieldStore {
    "actionsList": Array [],
    "fbEmmiter": BaseEventEmitter {
      "_currentSubscription": null,
      "_subscriber": EventSubscriptionVendor {
        "_currentSubscription": null,
        "_subscriptionsForType": Object {},
      },
    },
    "fieldId": "fake-field-id",
    "formId": "fake-form-id",
    "state": Object {
      "value": "",
    },
    "uniqueEventTypeNumber": 0,
  },
}
`;
